src/replica/replica.h

```diff
@@ -440,6 +440,12 @@ class replica : public serverlet<replica>, public ref_counter, public replica_ba
     void update_bool_envs(const std::map<std::string, std::string> &envs,
                           const std::string &name,
                           /*out*/ bool &value);
+    // get bool envs[name], return false if value is not bool
+    bool get_bool_envs(const std::map<std::string, std::string> &envs,
```

Comment:
these two functions have nothing to do with class replica. I think you needn't make them to be member functions for class replica

Reply:
I seperate `get_bool_envs` out of replica class, but remain `update_bool_envs`. It is actually update the filed of replica local variety, I think it's better to remain it in replica class.

Reply:
Which one does `update_bool_envs` update? Sorry I didn't find it.

Reply:
https://github.com/XiaoMi/rdsn/blob/982c2bd180a081fe2c772191cad0ea279e0e2685/src/replica/replica_config.cpp#L557-L566

`_deny_client_write` and `_validate_partition_hash`, although they are passed into `update_bool_envs`, but actually update the value of replica class.

Reply:
I don't agree with you. These two local values are passed by parameters. No need to make it a member function

Reply:
I still insist on my opinion now. In my view, function `update_bool_envs` is only a help function to simply the code to update the bool envs of replica, it just like function `update_throttle_env_internal`, it is not a utility function doesn't have any relationship with the replica class.

Reply:
I know what you mean. It looks strange to me that when looks into the implementation of this function, it updates nothing related to replica.

Reply:
And envs belongs to app_info, not replica

Reply:
> function update_bool_envs is only a help function to simply the code to update the bool envs of replica

The meaning of the sentence above is that it is the help function to simply the code to update `_deny_client_write` and `_validate_partition_hash` of replica class, `the bool envs of replica` is the simple statement.

> I know what you mean. It looks strange to me that when looks into the implementation of this function, it updates nothing related to replica.

If it move out of replica class, it will be like:
```
void update_bool_envs(const std::string &replica_name,
                                      const std::map<std::string, std::string> &envs,
                                      const std::string &name,
                                       bool &value)
{
    bool new_value = false;
    if (!get_bool_envs(envs, name, new_value)) {
        dwarn_f("{}: invalid value of env {}", replica_name, name);
        return;
    }
    if (new_value != value) {
        ddebug_f("{}, switch env[{}] from {} to {}", replica_name, name, value, new_value);
        value = new_value;
    }
}
```
I got your point, but it is also strange to move this function out of class replica. As the comment I said, it just like function `update_throttle_env_internal`, you can reference this function.